kit/Kit.cpp

```diff
@@ -2383,16 +2383,36 @@ void Document::drainQueue()
             }
         }
 
+        const size_t tilesQueued = _queue->getTileQueueSize();
         if (processInputEnabled() && !isLoadOngoing() &&
-            !isBackgroundSaveProcess() && _queue->getTileQueueSize() > 0)
+            !isBackgroundSaveProcess() && tilesQueued > 0)
         {
             std::vector<TileCombined> tileRequests = _queue->popWholeTileQueue();
-
-            // Put requests that include tiles in the visible area to the front to handle those first
-            std::partition(tileRequests.begin(), tileRequests.end(), [this](const TileCombined& req) {
-                return isTileRequestInsideVisibleArea(req); });
-            for (auto& tileCombined : tileRequests)
-                renderTiles(tileCombined);
+            std::vector<TileCombined*> earmarkedTiles;
+            size_t renderedTiles = 0;
+            for(TileCombined& tileCombined : tileRequests)
+            {
+                if (isIdle || (renderedTiles==0 && isTileRequestInsideVisibleArea(tileCombined)))
+                {
+                    renderTiles(tileCombined);
+                    renderedTiles += tileCombined.getTiles().size();
+                }
+                else if(renderedTiles>0)
+                    _queue->pushTileQueue(tileCombined.getTiles());
+                else
+                    earmarkedTiles.push_back(&tileCombined);
+            }
+            for(TileCombined* tileCombined : earmarkedTiles)
```

Comment:
yes, we could end up missing rendering tiles if not becoming idle, i.e. the ` unit-tiletest` (`TileCacheTests::testClientPartCalc`) failed otherwise waiting for one invisible tile.